### PR TITLE
o/assertstate: fix missing 'scheduled' header when auto refreshing assertions

### PR DESCRIPTION
**Pull request by MiguelPires**
When making requests to the snap store, some refresh requests with the 'fetch assertion' action were missing the "Snap-Refresh-Reason: scheduled" header. This was due to an unset RefreshOption when the assertions were auto refreshed. This PR sets the RefreshOptions when calling the store to make the refresh request.

https://bugs.launchpad.net/snapd/+bug/1937309

### CODE REVIEW

File: overlord/assertstate/bulk.go
Comment (mardy) [inline overlord/assertstate/bulk.go:245]: While in the other functions (which contain the word "refresh" in their name) it's quite obvious what the "auto" parameter does, here it isn't exactly so. Therefore I'd suggest renaming it to `autoRefresh`.

Comment (mardy) [reply overlord/assertstate/bulk.go:245]: Or `IsAutoRefresh`, I'm not sure which one is better :-)

File: overlord/assertstate/bulk.go
Comment (mardy) [inline overlord/assertstate/bulk.go:316]: same here about the parameter name

File: overlord/assertstate/bulk.go
Comment (mardy) [inline overlord/assertstate/bulk.go:178]: Using a variable would make it more clear what the last parameter is for:
```suggestion
	autoRefresh := true
	err := resolvePoolNoFallback(s, pool, beforeCommitChecker, userID, deviceCtx, autoRefresh)
```

**Review by mardy — COMMENTED** [review]
A couple of minor comments :-)

But it doesn't seem to me that the commit message is relevant to these changes, as it talks about some header which I cannot see in the diff.

**Comment by MiguelPires** [comment]
Thanks for the feedback @mardy :slightly_smiling_face:  I see your point, I'll amend the commit message to something like "Set missing RefreshOption when auto refreshing assertions" which is more relevant to the actual change

File: overlord/assertstate/bulk.go
Comment (mvo5) [reply overlord/assertstate/bulk.go:245]: I will comment on this some more later (in the middle of meetings right now) but I am not a big fan ob non-obvious booleans in API. I.e. "foo(..., true)" will not tell much about what the bool means. I would suggest to follow the pattern we use in other function and pass a "resolveOptions struct" (name TBD) in that can be nil (and then the defaults will be used). On that struct we initially only have a single member "IsAutoRefresh" (or something like this). And then it becomes: "foo(..., &resolveOptions{IsAutoRefresh: true\)"

File: .gitignore
Comment (mvo5) [inline .gitignore:10]: (nitpick^2) a bit unrelated to this PR but I understand that it's a bit silly to open a separate PR just for this change

File: overlord/assertstate/assertstate.go
Comment (mvo5) [inline overlord/assertstate/assertstate.go:60]: I will leave this to @pedronis  to decide so maybe best wait for his opinion on this: I personally would slightly prefer to have an options struct on RefreshSnapDecleartions() instead of having two functions. Something like (names are strawmans right now):
```go
type RefreshSnapDeclarationsOptions struct {
    IsAutoRefresh bool
}
func RefreshSnapDeclarations(s *state.State, userID int, opts *RefreshSnapDeclarationsOptions) error {
    if opts == nil { opts = &RefreshSnapDeclarationsOptions{} };
...
```
and then pass this opts down into refreshSnapDeclarations, bulkRefreshSnapDeclarations etc.

But again, I will leave it to Samuele to make the final call on this. I think we want some kind of struct even if we keep two {Auto,}RefreshSnapDeclartions because bools in function declarations are a bit hard to read often. I.e. something like "SetAutomatic(true|false)" is easy to understand but for "refreshSnapDecarations(s, userID, false)" it's not obvious from reading the code in isolation what the "false" stands for and one needs to go back to the declaration of the method etc.

File: overlord/assertstate/assertstate_test.go
Comment (mvo5) [inline overlord/assertstate/assertstate_test.go:968]: Thanks for adding this! 

As another nitpick, I think `c.Check()` is slightly better here and below but we have no hard rules for when to use Assert and Check, it's a bit mixed. Below there is a bunch more that the test could check even when the IsAutoRefresh is not set. In general the rule of thumb is that Assert() is useful if it does not make sense to test anything beyond the this point and Check when one detail about the test is not correct but it may make sense to continue. With assert() one has to fix issues one-by-one, i.e. run test, fix one assert, run again, fix next assert etc. With Check() multiple issues are visible in the same test run.

**Review by mvo5 — COMMENTED** [review]
Thank you for this PR. I have a few inline comments, suggestions and nitpicks. But overall this looks very nice.

**Comment by MiguelPires** [comment]
Thanks for the feedback @mvo5. I added an options struct to replace the less-readable bool in the API and changed the assertion to use Check.

**Review by mardy — APPROVED** [review]
Let's wait for one more approval, but this looks good to me. Thanks Miguel!

File: overlord/assertstate/assertstate.go
Comment (pedronis) [reply overlord/assertstate/assertstate.go]: I think what was chosen, single function with options is fine. Pawel can tweak this further it later if needed

File: overlord/assertstate/assertstate.go
Comment (pedronis) [inline overlord/assertstate/assertstate.go:58]: this should follow the function name and have a plural:

RefreshSnapDeclarationsOptions

**Review by pedronis — APPROVED** [review]
lgtm, one comment about the options struct name

**Comment by pedronis** [comment]
@mvo please re-review this as well

File: overlord/assertstate/assertstate.go
Comment (anonymouse64) [inline overlord/assertstate/assertstate.go:347]: do we have a test case for checking that opts is true in this codepath? I see we do for ValidationSetAssertionForMonitor, but I'm unclear if we do for this case

File: overlord/assertstate/bulk.go
Comment (anonymouse64) [inline overlord/assertstate/bulk.go:178]: same question do we check that this option is specifically set to true in any test?

**Review by anonymouse64 — APPROVED** [review]
overall looks good, one nitpick about import ordering and questions about the test coverage, seems like we should try to have checks that the opts.IsAutoRefresh is true in more places, no?

File: overlord/assertstate/bulk.go
Comment (pedronis) [reply overlord/assertstate/bulk.go:178]: I misread this, this should come from outside through the levels for symmetry and also because some callers in the end will not be from auto-refresh I think

File: overlord/assertstate/assertstate.go
Comment (pedronis) [inline overlord/assertstate/assertstate.go:351]: this should also take opts basically

**Review by pedronis — CHANGES_REQUESTED** [review]
I read things a bit too quickly it seems. The calls for validation-sets should also take options and respect them.

File: overlord/assertstate/assertstate.go
Comment (MiguelPires) [reply overlord/assertstate/assertstate.go:351]: Should I rename the options to RefreshAssertionsOptions since it's used in both kinds of assertions (snap declarations and validation sets)? Or is it better to have two different options for each type of call?

File: daemon/export_test.go
Comment (MiguelPires) [reply daemon/export_test.go:24]: Good to know, thanks! I'll re-order them

File: overlord/assertstate/assertstate.go
Comment (MiguelPires) [reply overlord/assertstate/assertstate.go:347]: Yes, I added an assertion to TestValidationSetAssertionsAutoRefresh. I can add two different tests to ensure that both RefreshSnapDeclarations and RefreshValidationSetAssertions propagate the options

**Comment by MiguelPires** [comment]
Thanks everyone for the feedback. I refactored the options so both the validation set and snap declaration assertions use it. I also added a few additional checks to some tests to insure both calls handle the opts correctly

**Comment by pedronis** [comment]
@MiguelPires this needs a master merge as the now landed ValidationSetAssertionForEnforce needs changes too

**Review by pedronis — COMMENTED** [review]
lgtm apart the need for a merge and fix

**Comment by MiguelPires** [comment]
@stolowski I fixed the merge conflicts. Please have a look when you have a moment

**Comment by MiguelPires** [comment]
@stolowski when you can, this PR is ready for a pass from you

**Review by pedronis — APPROVED** [review]
thanks

File: overlord/assertstate/assertstate.go
Comment (stolowski) [inline overlord/assertstate/assertstate.go:572]: Shouldn't we simply pass have `opts := &RefreshAssertionsOptions{IsAutoRefresh: true}` here (and leave ValidationSetAssertionForEnforce signature intact - not take opts)?

Comment (MiguelPires) [reply overlord/assertstate/assertstate.go:572]: The idea was just to make the API consistent. I think Samuele mentioned in an earlier comment that the functions should be consistent (but I may have misunderstood/misapplied the meaning)

Comment (stolowski) [reply overlord/assertstate/assertstate.go:572]: Sorry, maybe I wasn't clear; I mean ValidationSetAssertionForMonitor doesn't take opts, but ValidationSetAssertionForEnforce does take them, so they are not consistent in that regard. Can we keep ValidationSetAssertionForEnfore consistent with  ValidationSetAssertionForMonitor and simply hardcode opts we pass to RefreshValidationSetAssertions here? I guess the question is when do we want to pass autorefresh:true and false here from the outside? On automatic snap refreshes vs manual installs maybe? CC @pedronis ?

Comment (pedronis) [reply overlord/assertstate/assertstate.go:572]: sorry, I missed this skimming the last commit. We want:

```
opts := &RefreshAssertionsOptions{IsAutoRefresh: false}
```
(false!, not true)

here afaict.  We don't plan to use ValidationSetAssertionFor* in the context of auto refresh so they don't need to take options.

**Review by stolowski — CHANGES_REQUESTED** [review]
Looks fine overall, needs one change discussed in the comments with Samuele.

**Review by stolowski — APPROVED** [review]
Thank you!

**Comment by MiguelPires** [comment]
@mvo5 the tests failures are unrelated, can you merge this one as well please?

**Pull request merged**
